### PR TITLE
[AutoDiff upstream] Add derivative function witness/vtable entry SILGen.

### PR DESCRIPTION
**Pull request by dan-zheng**
`@differentiable` attribute on protocol requirements and non-final class
members now produces derivative function entries in witness tables and vtables.

This enables `witness_method` and `class_method` differentiation.

Existing type-checking rules:

- Witness declarations of `@differentiable` protocol requirements must have a
  `@differentiable` attribute with the same configuration (or a configuration
  with superset parameter indices).
  - Witness table derivative function entries are SILGen'd for `@differentiable`
    witness declarations.

- Class vtable derivative function entries are SILGen'd for non-final
  `@differentiable` class members.
  - These derivative entries can be overridden or inherited, just like other
    vtable entries.

Resolves TF-1212.

---

See `test/AutoDiff/SILGen/{witness_,v}table.swift` for examples.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@rjmccall @slavapestov: as SIL code owners, would you like to review this patch?

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test Windows platform

**Comment by dan-zheng** [comment]
@swift-ci Please test Windows platform

**Comment by dan-zheng** [comment]
All tests passed on a previous commit, so I'll just trigger smoke tests now.
@swift-ci Please smoke test

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
Merging to unblock progress. Happy to address feedback later!

**Pull request merged**
